Issue/refactor reader post details fragment part ii

ashiagr opened the pull request:
This PR converts `ReaderPostDetailFragment.ShowPostTask` (async) task to coroutines and extracts parts of logic to VM, usecases.

It
1. Converts `ShowPostTask` async task to coroutines (8018ec4, 56323cf)
2. Extracts get local reader post logic from `ShowPostTask.doInBackground` to usecase (7378d41, f7320cc)
3. Extracts fetch reader post logic from `ShowPostTask.onPostExecute` -> `requestPost`, `onRequestFailure` to usecase (33498f4, 29e5e0c)
4. Introduces `ErrorUiState` 8d2501a
5. Introduces `LoadingUiState` 9ffdac4
6. Extracts fetch post success/ failure result handling from `handleShowPostResult` to VM (89a9744, 56323cf, f430ff4)

_RESULT_: `ShowPostTask`, `requestPost`, `handleShowPostResult` are eliminated from the fragment with logic moved to VM and usecases.

### Review Instructions
 I've kept all these changes in a single PR as it might be easier to test all changes (get/fetch post + result handling) for `ShowPostTask` at one place. PR size has increased mainly due to introduction of new tests. I can split part 2 and 3 into a child PR if it helps in the review. Let me know.

Since there's no hurry to merge this PR, anyone relatively free can review it.

### Merge Instructions

- Make sure PR #14243 is merged to `issue/refactor-reader-post-details-fragment` branch
- Review this PR
- Remove `Not Ready For Merge` label
- Merge PR to `issue/refactor-reader-post-details-fragment` branch

### Testing Instructions

Make sure 
- Newly added tests in `ReaderGetPostUseCaseTest`, `ReaderFetchPostUseCaseTest`, `ReaderPostDetailViewModelTest` cover important scenarios affected in the refactoring.
- Reader post details ui, loading and error ui is shown as before prior to refactoring.

For manual testing, I followed below steps: 

**Test.1a Local Post Found**

- Launch app and open reader post (post details for posts on reader tab already exist in db)
- Note that post details are shown properly

**Test.1b Local Editor Pick Discover Post Found**
 
I wasn't able to find  "Editor Pick Discover Post" and couldn't test this manually. These tests should  cover the scenario
- `given editor pick discover post is found, when reader post is retrieved, discover source post returned`
-  `given non editor pick reader post is found, when reader post is retrieved, reader post is returned`

**Test.2a Local Post Not Found, Fetch Post Loading**

- Replace `val post = discoverEditorPost ?: readerPost` with `val post = null` in `ReaderGetPostUseCase`
- Launch app and open reader post
- Note that progress bar is shown

**Test.2a-i Local Post Not Found, Fetch Post Success**

Since posts on the reader tab are already present in the db, I tested this scenario by loading a related post which also triggers `vm.showPost`.
- Start afresh, note down a reader post with related post not clicked earlier (i.e. not present in the local db)
- Launch app and open above noted reader post
- Note that 
   - related post screen is shown 
   - loading is shown
   - loading is hidden after few secs
   - related post details are shown

**Test.2b Local Post Not Found, Fetch Post No Network
Test.2c Local Post Not Found, Fetch Post Post Not Found
Test.2d Local Post Not Found, Fetch Post Generic Failure**

- Replace `val post = discoverEditorPost ?: readerPost` with `val post = null` in `ReaderGetPostUseCase`
- Return appropriate result from `ReaderFetchPostUseCase`
- Launch app and open reader post
- Note that progress bar is shown followed by appropriate error ui

no network| post not found|generic
---|---|---
![no_network](https://user-images.githubusercontent.com/1405144/110908850-6838f280-8335-11eb-99ba-0c47e00eb6b1.png)|![post_not_found](https://user-images.githubusercontent.com/1405144/110908880-6f600080-8335-11eb-90fd-61732726a08f.png)|![unable_to_retrieve_post](https://user-images.githubusercontent.com/1405144/110909218-d8477880-8335-11eb-9c93-81fd410f532b.png)


**Test.3a Local Post Not Found, Fetch Post Not Authorised
Test.3b Local Post Not Found, Fetch Post Not Authorised Fallback
Test.3c Local Post Not Found, Fetch Post Not Authorised with SignIn
Test.3d Local Post Not Found, Fetch Post Not Authorised Fallback with SignIn**

- Replace `val post = discoverEditorPost ?: readerPost` with `val post = null` in `ReaderGetPostUseCase`
- Return `NotAuthorised` result from `ReaderFetchPostUseCase`
- Set `shouldOfferSignIn` to true or false in `ReaderPostDetailViewModel`
- Set `interceptedUri` to null/ not null  in `ReaderPostDetailViewModel`
- Launch app and open reader post
- Note that progress bar is shown followed by appropriate not authorised ui

not authorised|not authorised fallback|not authorised sign in|not authorised fallback sign in
---|---|---|---
![not authorised](https://user-images.githubusercontent.com/1405144/110908913-7be45900-8335-11eb-8079-08e8f28c9b55.png)| ![not_authorised_fallback](https://user-images.githubusercontent.com/1405144/110908931-80107680-8335-11eb-9341-f89630d30ecc.png)|![not_authorised_sign_in](https://user-images.githubusercontent.com/1405144/110908948-869eee00-8335-11eb-98a1-9d9a8496338b.png)| ![not_authorised_falbback_sign](https://user-images.githubusercontent.com/1405144/110908971-8bfc3880-8335-11eb-8508-dacff136bdf1.png)

**Test.4 Adhoc testing**

- Launch app and open reader post
- Perform random actions: like/ bookmark/open related posts etc
- Make sure nothing has broken
 
PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Refactor: Store/ access isFeed from VM

ashiagr committed:
Refactor: Introduce loading state

ashiagr committed:
Refactor: Extract get reader post from db logic to usecase

"doInBackgroundShowPost" method in fragment was used to get local reader post. This logic is extracted to a usecase and "doInBackgroundShowPost" method is removed.

ashiagr committed:
Tests: Add tests for reader get post usecase

ashiagr committed:
Tests: Add tests for reader fetch post use case

ashiagr committed:
Refactor: Handle fetch post error

ashiagr committed:
Refactor: Convert async task to coroutine inside fragment temporarily

ashiagr committed:
Refactor: Extract fetch reader post logic to usecase

"requestPost" method is used to
1. fetch post
2. manage progress bar visibility on post loading
3. trigger success handling (recalls showPost)
4. trigger error handling (onRequestFailure)

This commit only extracts fetching logic (1.) and returns appropriate response on success or failure using a usecase.

Loading/Success/Error handling will be done in the VM in the upcoming commits. "requestPost" method is retained in the fragment till that point.

ashiagr committed:
Refactor: Fetch post in VM using usecase

This commit checks for local post in the VM and if not found, fetches post using reader fetch post use case.

It also
1. Temporarily removes call to "handleShowPostResult" in the fragment and adds fetch post usecase result cases in the VM. These cases will be handled in the upcoming commits.
2. Extracts "isPostTaskRunning" flag handling in fragment to "AlreadyRunning" fetch post result case.

ashiagr committed:
Refactor: Remove redundant requestPost, handleShowPostResult methods from fragment

ashiagr committed:
Refactor: Introduce error ui state

ashiagr committed:
Fix misc warnings

ashiagr committed:
Refactor: Launch show post action in VM

Rename vm.onShowPost to updatePostDetailsUi
Rename vm.onShowPostStarted to onShowPost
Call updatePostDetailsUi when post is found on show post
Call onPostExecuteShowPost inside fragment on ui update

ashiagr committed:
Tests: Add/ update tests for show reader post

ashiagr committed:
Merge branch 'issue/refactor-reader-post-details-fragment-part-i' into issue/refactor-reader-post-details-fragment-part-ii

ashiagr committed:
Fix: Add braces to multiline if-else detekt warning + minor refactoring updateNotAuthorisedErrorState

ashiagr committed:
Fix: ktlnit

ashiagr committed:
Fix: Detekt baseline

ashiagr committed:
Fix: AlreadyRunning handling

Matches previous behaviour that hides error text view when request is running

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/usecases/ReaderFetchPostUseCase.kt:
```diff
@@ -0,0 +1,97 @@
+        const val STATUS_CODE_200 = 200
+        const val STATUS_CODE_401 = 401
+        const val STATUS_CODE_403 = 403
+        const val STATUS_CODE_404 = 404
```
Suggestion (💡): Instead of using these newly generated constants, what if you replace that with the `HttpURLConnection` constants, which are available for the whole project?

```
STATUS_CODE_200 -> HttpURLConnection.HTTP_OK
STATUS_CODE_401 -> HttpURLConnection.HTTP_UNAUTHORIZED
STATUS_CODE_403 -> HttpURLConnection.HTTP_FORBIDDEN
STATUS_CODE_404 -> HttpURLConnection.HTTP_NOT_FOUND
```

PS: Also, you might then want to update the same hard coded constants within the `ReaderPostDetailFragment` class.

ashiagr replied:
I should've thought about it 😅. In 08f897b.

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/viewmodels/ReaderPostDetailViewModel.kt:
```diff
@@ -133,6 +159,52 @@ class ReaderPostDetailViewModel @Inject constructor(
+                FetchReaderPostState.Failed.RequestFailed ->
+                    _uiState.value = ErrorUiState(UiStringRes(R.string.reader_err_get_post_generic))
+
+                FetchReaderPostState.Failed.NotAuthorised -> updateAndTrackNotAuthorisedErrorState()
```
1. Suggestion (💡): When an `and` is involved within a function name, I usually make sure to follow the order of the naming with the order of the called logic. For example, I would have named that as `trackAndUpdateNotAuthorisedErrorState()` instead.
2. Suggestion (💡): When an `and` is involved within a function name, I usually split the function into two inner functions or lines, to reduce the functions complexity and adhere to what the name function suggests, like:
```
    private fun trackAndUpdateNotAuthorisedErrorState() {
        trackNotAuthorisedState()
        _uiState.value = ErrorUiState(
                message = UiStringRes(getNotAuthorisedErrorMessageRes()),
                signInButtonVisibility = shouldOfferSignIn
        )
    }

    private fun getNotAuthorisedErrorMessageRes() = if (!shouldOfferSignIn) {
        if (interceptedUri == null) {
            R.string.reader_err_get_post_not_authorized
        } else {
            R.string.reader_err_get_post_not_authorized_fallback
        }
    } else {
        if (interceptedUri == null) {
            R.string.reader_err_get_post_not_authorized_signin
        } else {
            R.string.reader_err_get_post_not_authorized_signin_fallback
        }
    }
```

ashiagr replied:
Great suggestion, included in db48550.

ParaskP7 replied:
Nope, YOU are great! 🚀

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.kt:
```diff
@@ -371,26 +360,42 @@ class ReaderPostDetailFragment : ViewPagerFragment(),
+            uiHelpers.updateVisibility(progress_loading, it.loadingVisible)
+            when (it) {
+                is LoadingUiState -> Unit // Do Nothing
+
```
Minor (🔍): Consider removing the extra empty lines.

ashiagr replied:
Removed in 4b0bd83

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.kt:
```diff
@@ -371,26 +360,42 @@ class ReaderPostDetailFragment : ViewPagerFragment(),
         viewModel.navigationEvents.observe(viewLifecycleOwner, { it.applyIfNotHandled { handleNavigationEvent() } })
 
-        viewModel.start(isRelatedPost)
+        viewModel.start(isRelatedPost, savedInstanceState ?: arguments)
```
Suggestion (💡): Instead of passing a `bundle`, which is Android specific, maybe passing the individual `isFeed` and `interceptedUri` arguments is better? I know that this is already happening and we are used to passing a `bundle` to the view model, but given the fact that view models should be framework agnostic maybe it is better if we separate those?

PS: I know that this is an `ViewModel` specific to Android, but still the general suggestion is to avoid putting any Android specific code in it.

ashiagr replied:
I twice thought about it during refactoring, and noticed this existing convention so passed bundle for consistency. 
I agree it is better to avoid it, suggestion accepted in a959c8a.

ParaskP7 replied:
🙏

ParaskP7 commented on WordPress/src/test/java/org/wordpress/android/ui/reader/usecases/ReaderGetPostUseCaseTest.kt:
```diff
@@ -0,0 +1,111 @@
+class ReaderGetPostUseCaseTest : BaseUnitTest() {
+    @Mock private lateinit var readerPostTableWrapper: ReaderPostTableWrapper
+    private val postId = 1L
+    private val blogId = 2L
```
Suggestion (💡): What if those ids are named to:
```
    private val readerPostId = 1L
    private val readerBlogId = 2L
    private val discoverPostId = 3L
    private val discoverBlogId = 4L
```

And then used as:
```
    private val readerPost = ReaderPost().apply {
        postId = readerPostId
        blogId = readerBlogId
    }

    private val readerDiscoverSourcePost = ReaderPost().apply {
        postId = discoverPostId
        blogId = discoverBlogId
    }
```

Just to make it more visible what is what and avoid doing the `this@ReaderGetPostUseCaseTest` trick?

ashiagr replied:
Nice 👍, in 843e773

ParaskP7 commented on WordPress/src/test/java/org/wordpress/android/ui/reader/usecases/ReaderGetPostUseCaseTest.kt:
```diff
@@ -0,0 +1,111 @@
+                assertThat(result).isEqualTo(Pair<ReaderPost?, Boolean>(readerPost, false))
+            }
+
+    private fun createPost(
```
Minor (🔍): It seems that this helper test function doesn't need the `isDiscoverPost` parameter as this is always `true`. As such what if that is rewritten to:
```
    private fun createDiscoverPost(
        discoverType: DiscoverType = DiscoverType.OTHER
    ): ReaderPost {
        val post = spy(readerPost)
        // The ReaderPost contains business logic and accesses static classes. Using spy() allows us to use it in tests.
        whenever(post.isDiscoverPost).thenReturn(true)
        val mockedDiscoverData: ReaderPostDiscoverData = mock()
        whenever(post.discoverData).thenReturn(mockedDiscoverData)
        whenever(mockedDiscoverData.discoverType).thenReturn(discoverType)
        whenever(mockedDiscoverData.postId).thenReturn(readerDiscoverSourcePost.postId)
        whenever(mockedDiscoverData.blogId).thenReturn(readerDiscoverSourcePost.blogId)
        return post
    }
```

ashiagr replied:
Removed in f469d49

ParaskP7 reviewed (approved):
👋 @ashiagr  !

I have reviewed and tested this PR as per the instructions, everything works as expected, terrific job on both the refactoring you did and the tests you added! 🌟 🌟 🌟 

I have left only minor (🔍) comments and suggestions (💡) for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

PS: I couldn't do `Test.2a`, see 3rd bullet point. I am not sure if the `... followed by post details` is a typo or not, but following these instructions I was seeing a persistent `Loading` screen only. 
PSS: You update the Detekt baseline, kudos for that! When doing so, I usually suggest that no new issues are added to the baseline. Updates are okay, and especially removals are encouraged, but for new issues we should either try to fix them or suppress them locally in order to make them explicit and not letting the baseline grow. Wdyt?

ashiagr committed:
Fix: Replace newly added status constants with HttpURLConnection constants

ashiagr committed:
Fix: Rename method based on order of called logic, split into an inner fn

ashiagr committed:
Fix: Remove empty lines in when block

ashiagr committed:
Fix: Remove isDiscoverPost parameter from helper test function

ashiagr committed:
Fix: Pass constants directly to VM, avoid passing bundle

ashiagr committed:
Fix: Rename ids for better readability

ashiagr commented:
Thanks @ParaskP7 for a great review once again! Really appreciate your time 🙏.

> PS: I couldn't do Test.2a, see 3rd bullet point. I am not sure if the ... followed by post details is a typo or not, but following these instructions I was seeing a persistent Loading screen only.

I've updated `Test.2a` and added an additional `Test.2a-i`. Thanks for catching it 🙇‍♀️.

> PSS: You update the Detekt baseline, kudos for that! When doing so, I usually suggest that no new issues are added to the baseline. Updates are okay, and especially removals are encouraged, but for new issues we should either try to fix them or suppress them locally in order to make them explicit and not letting the baseline grow. Wdyt?

I agree and will follow this advice. I see only removals in eccc60d. Is that ok?

ashiagr committed:
Fix: Replace 200 with HttpURLConnection status code

ParaskP7 commented:
👋 @ashiagr !

> I've updated Test.2a and added an additional Test.2a-i. Thanks for catching it 🙇‍♀️.

Thank you so much! 🙏 

> I agree and will follow this advice. I see only removals in eccc60d. Is that ok?

Many thanks again! Yes eccc60d95c0ed6610543f7e10b8904ffe9fa6aa5 is just fine. This comment of mine was triggered because of the other commit, the 58a5d4ff653d37b26484bd02a964f59bef733be6. Not sure how it end up in my IDE version control log, but I can't find it within [commit list](https://github.com/wordpress-mobile/WordPress-Android/pull/14244/commits) in this PR... 🤷

ParaskP7 reviewed (approved):
I double checked all your changes, LGTM! Thank you so much for the effort, really really great work here! 🌟 

PS: Feel free to merge when CI is complete.

ashiagr commented:
> I believe there exists a 200 that you can replace with HttpURLConnection.HTTP_OK in this class as well.

Replaced in e210490

ParaskP7 commented:
👋 @ashiagr !

> Replaced in e210490

Yeap, I saw that already prior my previous comment, still approved! 🥇

ashiagr committed:
Merge branch 'issue/refactor-reader-post-details-fragment-part-i' into issue/refactor-reader-post-details-fragment-part-ii

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.kt
#	WordPress/src/main/java/org/wordpress/android/ui/reader/discover/viewholders/ReaderRelatedPostViewHolder.kt

ashiagr committed:
Fix merge
